[CIR][CIRGen] Add CIRGen support for float16 and bfloat

Lancern opened the pull request:
This PR adds two new CIR floating-point types, namely `!cir.f16` and `!cir.bf16`, to represent the float16 format and bfloat format, respectively.

This PR converts the clang extension type `_Float16` to `!cir.f16`, and converts the clang extension type `__bf16` type to `!cir.bf16`. The type conversion for clang extension type `__fp16` is not included in this PR since it requires additional work during CIRGen.

Only CIRGen is implemented here, LLVMIR lowering / MLIR lowering should come next.

bcardosolopes commented:
Nathan just did a rebase, can you please update? Sorry for the churn

Lancern commented:
Rebased onto the latest `main`.

bcardosolopes commented on clang/lib/CIR/CodeGen/CIRGenTypeCache.h:
```diff
@@ -35,9 +35,10 @@ struct CIRGenTypeCache {
   /// half, bfloat, float, double
-  // mlir::Type HalfTy, BFloatTy;
   // TODO(cir): perhaps we should abstract long double variations into a custom
   // cir.long_double type. Said type would also hold the semantics for lowering.
```
Looks like this comment isn't relevant anymore? You might as well delete it already

Lancern replied:
Updated.

bcardosolopes commented on clang/include/clang/CIR/Dialect/IR/CIRTypes.td:
```diff
@@ -179,7 +193,8 @@ def CIR_LongDouble : CIR_FloatType<"LongDouble", "long_double"> {
 // Constraints
 
-def CIR_AnyFloat: AnyTypeOf<[CIR_Single, CIR_Double, CIR_LongDouble]>;
+def CIR_AnyFloat: AnyTypeOf<[CIR_Single, CIR_Double, CIR_FP16, CIR_BFloat16, 
```
Do all operations currently using `CIR_AnyFloat` support 16-bit fp?

Lancern replied:
I believe yes. The operations currently using `CIR_AnyFloat` include:

- Builtin fp2fp unary operations family, e.g. `cir.ceil`, `cir.cos`, `cir.exp`, etc.
- Atomic fetch operation.

All operations above should support `f16` and `bf16`. See https://godbolt.org/z/rq5edMarM .

bcardosolopes commented on clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp:
```diff
@@ -3021,6 +3021,12 @@ void prepareTypeConverter(mlir::LLVMTypeConverter &converter,
     return mlir::FloatType::getF64(type.getContext());
   });
+  converter.addConversion([&](mlir::cir::FP16Type type) -> mlir::Type {
+    return mlir::FloatType::getF16(type.getContext());
```
I'd like to see LLVM lowering tests for all these new types.

Lancern replied:
The LLVM lowering for the types themselves is covered in [clang/test/CIR/Lowering/float.cir](https://github.com/llvm/clangir/pull/571/files#diff-64dcb563e6b26fa528a040c561b53ad425ab85891a369a425f04418134091082) .

bcardosolopes commented on clang/test/CIR/CodeGen/float16-arithmetic.c:
```diff
@@ -0,0 +1,99 @@
+
+_Float16 add1(_Float16 a, _Float16 b) {
+  return a + b;
+}
```
16-bit fp requires different LLVM lowering for different arches, example: ARM64 vs x86_64. I'm a bit skeptical this PR will just automatically work properly (I don't see any arch specific checks, and perhaps there's some ABI work missing?), quick examples:
x86_64: https://godbolt.org/z/vr4hG6v3x
ARM64: https://godbolt.org/z/Mn8znbbhx

Handwaving to proper support right now might give us the wrong impression that this actually work, where in truth it doesn't and will require folks to spend hours investigating why it does not. What I'd like to see here:
- LLVM checking for one at least one Arch.
- Should some of this arch specific things be handled in CIR already? I'd expect this to already be the case for other floating point types, which makes me worried that current fp support took the shortest path and might not actually be right when we lower down to LLVM (e.g. fp80) - I hope I'm wrong cause that's gonna be hard to pinpoint when we get correctness issues while running code.
- Somehow make sure that LowerToLLVM.cpp (or CIRGen if details get exposed to that level) gives us error/asserts for arches it doesn't know how to do the right job.

Lancern replied:
I checked the upstream clang CodeGen code. The relevant code is:

https://github.com/llvm/clangir/blob/981c976f692c1618714348967ddaca55e99ce76e/clang/lib/CodeGen/CGExprScalar.cpp#L841-L847

The basic idea is, before emitting a unary or binary scalar expression of floating point type, CodeGen checks whether the operands should be promoted by calling `Ty.UseExcessPrecision`. If this call returns `true`, the operands are promoted to `float` before evaluating the expression. The call is target-specific, on x86 it returns `true` and on `aarch64` it returns `false`. So the promotion of the operands need to be handled similarly at the CIR side if we're going to align to the upstream clang CodeGen. Here is a possible approach:

- Invent a new attribute that holds global floating point related configurations (such as whether half-precision types should be promoted), and attach this attribute to the `module` operation, just like `#cir.lang` and `#cir.sob`.
- During LLVM lowering prepare, teach the transformation pass to promote the operands of `cir.unary` and `cir.binop` accordingly from the information given by the new module attribute.

This may be a non-trivial change. Any thoughts on this?

bcardosolopes replied:
Thanks for pointing out the difference in terms of code. Seems like some of my previous PR reviews slip through the cracks on letting these operations to be emitted without proper calls to `getPromotionType(...);`.

I'm a bit confused on why do you think this needs to be done post CIRGen - we need to do these promotions/extensions during CIRGen (by calling `getPromotionType`, etc), I don't see any current advantage of postponing these specific ABI lowerings.

I'm coming from the correctness angle here, if things like this sneak in without proper correctness, it will be a lot of work to investigate these issues during runtime later on. I appreciate your work on adding these fp variants but we first need to make sure that we have a sound path (or at least enough asserts for things we don't yet handle properly).

So, it's not because there's a problem with the status quo that we should ignore it and keep moving on. What I'd like to see before this is ready, probably in some other PR, is some work on making sure the skeleton of these ops get closer to OG codegen and that we handle these promotions/extensions properly during CIRGen - this doesn't mean you need to implement logic for every target or ABI, but the usual way should apply: similar skeleton and asserts for what's not done - anything is better than silently miscompilation. With that in, this PR can then land with checks that use the proper target triple to check their behaviors.

bcardosolopes reviewed (changes requested):
Thanks for adding more floating point goodness. Comments inline.

Lancern commented:
Rebased onto the latest `main` and updated this PR according to the latest reviews:

- This PR now only deals with the CIRGen of the two new floating point types as this is already complex enough to be stuffed into a single PR.
- `CIRGenExprScalar` is updated according to the original clang CodeGen. Specifically, scalar promotion of the new half-precision types is added to the CIRGen.

bcardosolopes commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -478,14 +488,38 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
-      auto kind =
-          (isInc ? mlir::cir::UnaryOpKind::Inc : mlir::cir::UnaryOpKind::Dec);
-      value = buildUnaryOp(E, kind, input);
+      if (type->isHalfType())
```
In general, whenever following skeleton, I advice to stay closer as possible structurally. Things missing:
- assert+unimplemented for CGFPOptionsRAII
- `&& !CGF.getContext().getLangOpts().NativeHalfType` here

Lancern replied:
Updated.

bcardosolopes commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -478,14 +488,38 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
     } else if (type->isVectorType()) {
       llvm_unreachable("no vector inc/dec yet");
     } else if (type->isRealFloatingType()) {
```
Thanks for addressing most of the bits!

bcardosolopes commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -478,14 +488,38 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+        auto amt = Builder.getConstant(
+            loc, mlir::cir::FPAttr::get(value.getType(), F));
+        value = Builder.createBinop(value, mlir::cir::BinOpKind::Add, amt);
+      }
```
Another unimplemented unrecheable missing for if `(type->isHalfType() && !CGF.getContext().getLangOpts().NativeHalfType) {`

Lancern replied:
Updated.

bcardosolopes commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -478,14 +488,38 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+        // Create the inc/dec operation.
+        auto kind =
+            (isInc ? mlir::cir::UnaryOpKind::Inc : mlir::cir::UnaryOpKind::Dec);
+        value = buildUnaryOp(E, kind, input);
```
I'm a bit curious about the assumption of using unary ops here. I forgot about this and had to look above for `// NOTE(CIR): clang calls CreateAdd but folds this to a unary op`. All it takes for LLVM to fold `fadd` to unary is the value being single or double? Seems like an early optimization. Since this is the status quo for other handling that it's not specific to this patch, no worries to change it right now, but I'll like a comment here explaining why this is using a unary op and how this assumption works.

Lancern replied:
Updated.

bcardosolopes commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -752,13 +787,17 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+    if (const auto *VecType = dyn_cast_or_null<VectorType>(Result.FullType)) {
       Result.CompType = VecType->getElementType();
     }
     Result.Opcode = E->getOpcode();
```
While here, can you plz add an unimplented for `getFPFeaturesInEffect` below?

Lancern replied:
Updated.

bcardosolopes commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -793,15 +832,22 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
 #define HANDLEBINOP(OP)                                                        \
   mlir::Value VisitBin##OP(const BinaryOperator *E) {                          \
-    return build##OP(buildBinOps(E));                                          \
+    QualType promotionTy = getPromotionType(E->getType());                     \
```
Very nice!

bcardosolopes commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -1938,6 +2007,44 @@ mlir::Value ScalarExprEmitter::buildNullValue(QualType Ty, mlir::Location loc) {
 
+mlir::Value ScalarExprEmitter::buildPromoted(const Expr *E,
+                                             QualType PromotionType) {
+  E = E->IgnoreParens();
```
Very nice too!

bcardosolopes reviewed (changes requested):
Great, thanks for all the promotion related improvements, they are very nice.

> LLVMIR lowering, and MLIR lowering for the two new types are also included in this PR.

Fine for this, but for next PRs I might start getting more picky about adding LLVM IR checks, the unary questions below are one example where checking intended LLVM output could have made some of the choices easier.
